### ISSUE TITLE
Detection latency when start microphone stream

### ISSUE THREAD

**Issue by jen-co** [alphacep/vosk-api#219]
Hello,

Thank you for this excellent library and documentation. 

I am using the python api to process real time audio which I am streaming from a microphone input (using PyAudio). I have noticed that when I first start processing the audio there is a significant amount of time (about 8 seconds) where I am speaking but the partial result and final result are completely blank. Thereafter, the detection latency and accuracy are working really well. Is there anyway to reduce this initial latency?

For extra clarity I am using pyaudio input in non-blocking callback mode. I have set sample rate to 16000 Hz and frames per buffer to 4000:

stream = p.open(format=pyaudio.paInt16, channels=1, rate=16000, input=True, stream_callback=callback, frames_per_buffer=4000)

I did try open the microphone stream for a while before I actually process the audio using vosk but that didnt seem to make a difference to the latency.

Thanks,
Jennifer

**Comment by nshmyrev** [comment]
Hi! Thank you for trying Vosk. 

What is your hardware please? OS? Microphone?

**Comment by jen-co** [comment]
Hardware - Intel NUC10iFNK running Windows 10. Microphone is Sharkoon SM1 Notebook microphone (https://www.amazon.co.uk/Sharkoon-SM1-Notebook-microphone-Wired/dp/B077GCGC9K)

**Comment by nshmyrev** [comment]
Hm, ok.

What if you simply read the audio without feeding it into vosk, is there a latency? You can add print in callback to figure out.

Also, can you dump recorded audio to file and share?

**Comment by jen-co** [comment]
Yes. I have tried only reading the audio and not feeding it to vosk for at least 8 seconds. Thereafter I feed into vosk but there still seems to be the same initial detection latency.  Additionally, I am checking the audio amplitude the whole time and its appears to be peaking in the same manner when I am talking throughout so it doesn't seem like anything is off with the input audio. However, to be extra sure I will do a recording of the input audio today and share asap. Thanks again for your quick responses and help so far!

**Comment by nshmyrev** [comment]
Ok.

> Yes. I have tried only reading the audio and not feeding it to vosk for at least 8 seconds. Thereafter I feed into vosk but there still seems to be the same initial detection latency.

The question is not about vosk. How soon the callback audio appears after you start streaming? Is there a delay?

**Comment by jen-co** [comment]
I am doing the vosk processing in the audio callback and printing the output there, so it it is definitely getting the audio and processing it when I expect it to. I can post my code in a bit to explain this better.

**Comment by nshmyrev** [comment]
> I can post my code in a bit to explain this better.

I understand your approach already. Again, the delay may be in your audio subsystem, not in vosk. To verify that you need to measure the delay before you start streaming and the first sample arrived in callback. To do that you can add logging prints with timestamps.

**Comment by jen-co** [comment]
But surely if the callback is getting fired we know the audio is already there? Or are you saying that eventhough the callback has fired we can't be sure that actual valid audio is present?

In any case here is the code:

```from vosk import Model, KaldiRecognizer
from backend.utils import resource_path
import pyaudio
import os
import json


class SpeechDetection:
    def __init__(self, debug=True):
        self.debug = debug
        model = Model(resource_path(os.path.join("data", "speech-model")))
        self.fs = 16000
        self.rec = KaldiRecognizer(model, self.fs)
        self.p = pyaudio.PyAudio()
        self.chunksz = 4000
        self.stream = None
        self.result = None
        self.is_recording = False

    def start(self):
        if self.is_recording is False:
            print("start listening")
            self.is_recording = True
            self.stream = self.p.open(format=pyaudio.paInt16, channels=1, rate=self.fs, input=True,
                                      stream_callback=self.callback, frames_per_buffer=self.chunksz)
            self.stream.start_stream()

    def stop(self):
        print("stop listening")
        self.result = None
        if self.stream is not None:
            self.stream.stop_stream()
            self.stream.close()
            self.is_recording = False

    def is_running(self):
        return self.is_recording

    def close(self):
        self.p.terminate()

    def callback(self, in_data, frame_count, time_info, status):

        if self.rec.AcceptWaveform(in_data):
            result = json.loads(self.rec.Result())
            print(result)
        else:
            result = json.loads(self.rec.PartialResult())
            print(result)

        return in_data, pyaudio.paContinue


if __name__ == "__main__":
    speechdet = SpeechDetection()
    speechdet.start()

    while True:
        pass```

**Comment by jen-co** [comment]
(aarg sorry I dont know why github is not formatting the code snippet properly :/)

**Comment by nshmyrev** [comment]
> (aarg sorry I dont know why github is not formatting the code snippet properly :/)

You need to use triple ticks:

```
code
```

**Comment by nshmyrev** [comment]
> But surely if the callback is getting fired we know the audio is already there?

Yes, audio is there. But again, I want to know the delay between this:

```
            self.stream.start_stream()
```

and this:

```
    def callback(self, in_data, frame_count, time_info, status):
```

**Comment by jen-co** [comment]
The delay from start stream to callback is about 0.2 seconds, which is the same as the delay between each subsequent callback. 

I think we may still be misunderstanding each other, and apologies for this but in case I will clarify further what the problem is. For the first 8 seconds or so when vosk is processing the audio, eventhough I am speaking - partial result is a blank string and the final result either doesnt come or if it does it is quite inaccurate. 

So for example here I start processing and I say the word "testing" repeatedly at the same volume level and only after a while does it actually seem to detect the word "testing".  Here is an example of what it prints out:

start listening
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'result': [{'conf': 0.931092, 'end': 1.44, 'start': 1.23, 'word': 'yes'}], 'text': 'yes'}
{'partial': ''}
{'partial': ''}
{'partial': 'just'}
{'partial': 'testing'}
{'partial': 'testing'}
{'result': [{'conf': 1.0, 'end': 5.94, 'start': 5.4, 'word': 'testing'}], 'text': 'testing'}

**Comment by nshmyrev** [comment]
> For the first 8 seconds or so when vosk is processing the audio, eventhough I am speaking - partial result is a blank string

Ok, great, that clarifies the delay a bit.

Now can you please dump the audio data you pass into vosk into a file and share?

**Comment by jen-co** [comment]
Attached the audio file here
[test3.zip](https://github.com/alphacep/vosk-api/files/5244838/test3.zip)

And below is the print out of detection while this was streaming (as you can hear from the sample I am saying testing about 7 times from 1 second in but it only seems to pick up the last two instances of testing. From there on it is very responsive and accurate).

start listening
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'result': [{'conf': 1.0, 'end': 4.14, 'start': 3.9, 'word': 'mhm'}], 'text': 'mhm'}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': 'just'}
{'partial': 'just'}
{'partial': 'just'}
{'partial': 'just to'}
{'partial': 'just just'}
{'partial': 'just testing'}
{'partial': 'just testing'}
{'partial': 'just testing'}
{'partial': 'testing testing'}
{'partial': 'testing testing'}
{'partial': 'testing testing'}
{'partial': 'testing testing'}
{'result': [{'conf': 1.0, 'end': 7.32, 'start': 6.66, 'word': 'testing'}, {'conf': 1.0, 'end': 8.37, 'start': 7.77, 'word': 'testing'}], 'text': 'testing testing'}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'text': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': ''}
{'partial': "just didn't want"}
{'partial': "just didn't want to"}
{'partial': 'testing one two'}
{'partial': 'testing one two three'}
{'partial': 'testing one two three'}

**Comment by nshmyrev** [comment]
> Attached the audio file here test3.zip

Great, the click on the start is suspicious

Can you dump the same audio without running vosk? Just dump the audio from callback and don't feed it into the recognizer.

**Comment by jen-co** [comment]
Here is an audio sample without running vosk. There is still a click sound on start. 
[test4.zip](https://github.com/alphacep/vosk-api/files/5244913/test4.zip)

Also, for sanity check I ran the test3.wav sample on the command line using kaldi's online2-wav-nnet3-latgen-faster and it picked up all seven testing utterances. Next I will test it on the python vosk api with the loaded wav file (as opposed to real time streaming) to see if that is consistent.

**Comment by nshmyrev** [comment]
> Here is an audio sample without running vosk. There is still a click sound on start.
test4.zip

Good. I wonder what happens if you record with pyaudio without callback. Is there a click still?

Overall I am not very happy with pyaudio, it is broken in many ways.

**Comment by jen-co** [comment]
Ok another update, when I run the vosk python api for just loading the wave file it also fails to detect the first few "testing" utterances so it the problem does seem to be in the vosk python api.

If the click were an issue though, why do I get the correct results when I run the wav file in Kaldi's commandline tool?

I will try without callback in a bit and see if that helps. May have to be later today though as I have to get onto some other work. Thanks so much for all the help so far!

**Comment by nshmyrev** [comment]
> If the click were an issue though, why do I get the correct results when I run the wav file in Kaldi's commandline tool?

Ok, let me know if you are running online2-wav-nnet3-latgen-faster with the same model or with a different model.

**Comment by jen-co** [comment]
ok some more updates:

One thing I should have mentioned from the start is that I am using a custom model that I customized according to [this ](https://chrisearch.wordpress.com/2017/03/11/speech-recognition-using-kaldi-extending-and-using-the-aspire-model/) tutorial

When I run the online2-wav-nnet3-latgen-faster on this custom model it works accurately (i.e. all words are detected).

I have now tested 3 different models using the python vosk api processing test3.wav and these are the results:

- On [vosk-model-small-en-us-0.3](http://alphacephei.com/vosk/models/vosk-model-small-en-us-0.3.zip): processing test3.wav picks up all 7 "testing" utterances.

- On [vosk-model-en-us-aspire-0.2](https://alphacephei.com/vosk/models/vosk-model-en-us-aspire-0.2.zip): processing test3.wav does NOT pick up the initial "testing" utterances.

- On my custom model that is based off of the vosk-model-en-us-aspire-0.2 model I observe the same as above - i.e. processing test3.wav does NOT pick up the initial "testing" utterances.

So I guess the conclusion must be the somewhere along the lines the aspire model is not working correctly with the python api. Next thing I may try is to customise the small model and see what results I get from that.

**Comment by nshmyrev** [comment]
For best accuracy you need to use daanzu model between

**Comment by jen-co** [comment]
Hello again,

I was hoping it would be straight forward to try customise the daanzu model in the same way I did for the aspire model but I am not even sure where to start as I was hoping there would be folder in the kaldi/egs directory that would have the same format as the aspire one. This is not not the case.

Seeing that the aspire model does indeed work with online2-wav-nnet3-latgen-faster I tried to see if there were any obvious differences between the config passed to the the online2-wav-nnet3-latgen-faster command line and what is set up when using the vosk python api but so far nothing seems to help. Any ideas what else could be causing this discrepancy (i.e. that it works with online2-wav-nnet3-latgen-faster but not the python api)?

Thanks,
Jennifer

**Comment by nshmyrev** [comment]
Please share exact command lines and full outputs so I can reproduce.

**Comment by nshmyrev** [comment]
> I was hoping there would be folder in the kaldi/egs directory that would have the same format as the aspire one

You can get it here:

https://alphacephei.com/kaldi/models/vosk-model-en-us-daanzu-20200905-train.zip

**Comment by jen-co** [comment]
Here is what i pass to command line (for the original aspire model):

```
online2-wav-nnet3-latgen-faster \
  --online=false \
  --do-endpointing=false \
  --frame-subsampling-factor=3 \
  --config=exp/tdnn_7b_chain_online/conf/online.conf \
  --max-active=7000 \
  --beam=15.0 \
  --lattice-beam=6.0 \
  --acoustic-scale=1.0 \
  --word-symbol-table=exp/tdnn_7b_chain_online/graph_pp/words.txt \
  exp/tdnn_7b_chain_online/final.mdl \
  exp/tdnn_7b_chain_online/graph_pp/HCLG.fst \
  'ark:echo utterance-id1 utterance-id1|' \
  'scp:echo utterance-id1 test3.wav|' \
  'ark:/dev/null'
```

and here is the output:

```
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:ComputeDerivedVars():ivector-extractor.cc:183) Computing derived variables for iVector extractor
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:ComputeDerivedVars():ivector-extractor.cc:204) Done.
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:RemoveOrphanNodes():nnet-nnet.cc:948) Removed 1 orphan nodes.
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:RemoveOrphanComponents():nnet-nnet.cc:847) Removing 2 orphan components.
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:Collapse():nnet-utils.cc:1472) Added 1 components, removed 2
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:CompileLooped():nnet-compile-looped.cc:345) Spent 0.0177829 seconds in looped compilation.
utterance-id1 testing testing testing testing testing testing testing testing one two three 
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:main():online2-wav-nnet3-latgen-faster.cc:296) Decoded utterance utterance-id1
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:Print():online-timing.cc:55) Timing stats: real-time factor for offline decoding was 0.265571 = 4.71389 seconds  / 17.75 seconds.
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:main():online2-wav-nnet3-latgen-faster.cc:302) Decoded 1 utterances, 0 with errors.
LOG (online2-wav-nnet3-latgen-faster[5.5.746~2-8df25]:main():online2-wav-nnet3-latgen-faster.cc:304) Overall likelihood per frame was 1.74004 per frame over 591 frames.
```

Thanks for the daanzu link. Will try and customise it!

**Comment by nshmyrev** [comment]
Hi. We have reworked models and microphone code too:

https://github.com/alphacep/vosk-api/commit/6f2d6d0d69faacf3c9183818635cecca18ad6a97

can you please try and close if it is working

**Comment by jen-co** [comment]
Hi.

Thank you for the update. I will test this asap and let you know.

**Comment by jen-co** [comment]
Hi,

I have tested the microphone update. These are my findings:

1. There is no detection latency with the [vosk-model-small-en-us-0.15](http://alphacephei.com/vosk/models/vosk-model-small-en-us-0.15.zip) when I set the sample rate to 16000 (if I don't manually set the sample rate the code defaults to the device sample rate, which in this case is 44100 then the detection doesnt work at all.)

2. There is still a detection latency with the aspire model, which makes sense based on my later conclusions on this issue - i.e. that the issue lies somewhere in the way the vosk python api is processing the aspire model because the issue was not apparent when using the command line online2-wav-nnet3-latgen-faster with the aspire model.

In the end I was not able to customize the daanzu model in the same way that I did the aspire model as it wasnt clear from [this](https://chrisearch.wordpress.com/2017/03/11/speech-recognition-using-kaldi-extending-and-using-the-aspire-model/) tutorial how to do this. So for my application I just ensured that vosk was running the microphone input for a while before it was available to the user. 

This work around was acceptable in the end but I am hesitant to say the underlying issue has been resolved.
